File: c/avoid_flag_arguments.c
// Avoid Flag Arguments
// Avoid providing boolean arguments (true/false) to functions or methods.
// You could pass a string with a clearer name (for example).

#include <stdio.h>
#include <stdbool.h>
#include <string.h>

// Before
void greet_before(const char* name, bool formal) {
    if (formal) {
        printf("Good evening, %s.\n", name);
    } else {
        printf("Hi, %s!\n", name);
    }
}

// After
void greet(const char* name, const char* formality) {
    if (strcmp(formality, "formal") == 0) {
        printf("Good evening, %s.\n", name);
    } else if (strcmp(formality, "informal") == 0) {
        printf("Hi, %s!\n", name);
    }
}

int main() {
    greet_before("Alice", true);
    greet_before("Bob", false);

    printf("\n");

    greet("Alice", "formal");
    greet("Bob", "informal");

    return 0;
}

File: rust/avoid_flag_arguments.rs
// Avoid Flag Arguments
// Avoid providing boolean arguments (true/false) to functions or methods.
// You could pass a string with a clearer name (for example).

// Before
fn greet_before(name: &str, formal: bool) {
    if formal {
        println!("Good evening, {}.", name);
    } else {
        println!("Hi, {}!", name);
    }
}

// After
fn greet(name: &str, formality: &str) {
    match formality {
        "formal" => println!("Good evening, {}.", name),
        "informal" => println!("Hi, {}!", name),
        _ => {}
    }
}

fn main() {
    greet_before("Alice", true);
    greet_before("Bob", false);

    println!();

    greet("Alice", "formal");
    greet("Bob", "informal");
}
